pandas/core/ops.py

```diff
@@ -821,54 +816,72 @@ def na_op(x, y):
 
         return result
 
-    def wrapper(self, other, axis=None):
-        # Validate the axis parameter
-        if axis is not None:
-            self._get_axis_number(axis)
-
-        if isinstance(other, ABCSeries):
-            name = _maybe_match_name(self, other)
-            if not self._indexed_same(other):
-                msg = 'Can only compare identically-labeled Series objects'
-                raise ValueError(msg)
-            return self._constructor(na_op(self.values, other.values),
-                                     index=self.index, name=name)
-        elif isinstance(other, ABCDataFrame):  # pragma: no cover
+    def wrapper(self, other):
+        if isinstance(other, ABCDataFrame):  # pragma: no cover
             return NotImplemented
-        elif isinstance(other, (np.ndarray, pd.Index)):
-            # do not check length of zerodim array
-            # as it will broadcast
-            if (not is_scalar(lib.item_from_zerodim(other)) and
-                    len(self) != len(other)):
-                raise ValueError('Lengths must match to compare')
 
-            if isinstance(other, ABCPeriodIndex):
-                # temp workaround until fixing GH 13637
-                # tested in test_nat_comparisons
-                # (pandas.tests.series.test_operators.TestSeriesOperators)
-                return self._constructor(na_op(self.values,
-                                               other.astype(object).values),
-                                         index=self.index)
+        elif isinstance(other, ABCSeries) and not self._indexed_same(other):
+            msg = 'Can only compare identically-labeled Series objects'
+            raise ValueError(msg)
 
-            return self._constructor(na_op(self.values, np.asarray(other)),
-                                     index=self.index).__finalize__(self)
+        res_name = _get_series_op_result_name(self, other)
 
-        elif isinstance(other, pd.Categorical):
-            if not is_categorical_dtype(self):
-                msg = ("Cannot compare a Categorical for op {op} with Series "
-                       "of dtype {typ}.\nIf you want to compare values, use "
-                       "'series <op> np.asarray(other)'.")
-                raise TypeError(msg.format(op=op, typ=self.dtype))
+        if is_timedelta64_dtype(self):
+            res = op(pd.TimedeltaIndex(self), other)
+            return self._constructor(res, index=self.index, name=res_name)
+
+        elif is_datetime64_dtype(self) or is_datetime64tz_dtype(self):
+            # kludge; DatetimeIndex refuses to compare against None or
+            # datetime.date; until the "right" behavior is resolved, we cast
+            # these types here to types that DatetimeIndex understand.
+            if type(other) is datetime.date:
+                other = datetime.datetime(other.year, other.month, other.day)
+            elif other is None:
```

Comment:
why is this None check needed? do we not do this in the op comparison directly?

Reply:
See above.

Reply:
and I don't like this code path at all, meaning you are special casing things, which IMHO should be handle inside the DTI comparisons.

Reply:
Yes; the next step is to put this on the back-burner and fix the DTI comparisons.